src/java.base/share/classes/java/util/ImmutableCollections.java

```diff
@@ -748,6 +768,83 @@ public int lastIndexOf(Object o) {
         }
     }
 
+    @jdk.internal.ValueBased
+    static final class StableList<E> extends AbstractImmutableList<E> {
+
+        @Stable
+        private final IntFunction<? extends E> mapper;
+        @Stable
+        private final StableValueImpl<E>[] backing;
```

Comment:
You can use a backing `@Stable Object[]` instead. It will reduce indirection when accessing this list.

Reply:
Can you please elaborate a bit more on your proposal @merykitty?

Reply:
What would be the difference between `@Stable StableValueImpl<E>[] backing` and `@Stable Object[] backing`?

Reply:
For an `Object[]`, you only need to load the object from the array and it is probably what you need. For a `StableValueImpl[]`, you need to load the `StableValueImpl` from the array, and load the value from that `StableValueImpl`, which is 2 levels of indirections.

Reply:
It's true that the storage can be flatter here -- that said, this can also be done as a later refactoring. One advantage of doing things the way @minborg did it here, is that it's fairly easy to prove that the code below is correct -- which makes the initial review easier.

Reply:
Ahh. Now I see what you mean. This is something we did in a handful of prototypes we explored. While it is true that there will be one indirection less, the complexity of the code is going to grow. Also, if the element is constant folded, it does not matter.

Reply:
We will explore this option after the first integration of Stable Values.